[JSC] Careful optimization of eval

Constellation opened the pull request:
#### d2420b690182fc40009f2049e0930d06f9b5b4ea
<pre>
[JSC] Careful optimization of eval
<a href="https://bugs.webkit.org/show_bug.cgi?id=293474">https://bugs.webkit.org/show_bug.cgi?id=293474</a>
<a href="https://rdar.apple.com/151902736">rdar://151902736</a>

Reviewed by Yijia Huang.

This patch carefully restructures executeEval code.

1. Avoid doing some unnecessary things when we do not need to
   instantiate any variables.
2. Use vmEntryToJavaScriptWith0Arguments if necessary.
3. Use ensureStillAliveHere to keep JITCode alive instead of ref-ing.
   This ThreadSafeRefCount turned out super costly.

* Source/JavaScriptCore/bytecode/CodeBlock.h:
(JSC::CodeBlock::functionDecls):
(JSC::CodeBlock::functionExprs):
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.h:
(JSC::UnlinkedCodeBlock::functionDecls const):
(JSC::UnlinkedCodeBlock::functionExprs const):
* Source/JavaScriptCore/bytecode/UnlinkedEvalCodeBlock.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeEval):
* Source/JavaScriptCore/runtime/EvalExecutable.h:
(JSC::EvalExecutable::variables const):
(JSC::EvalExecutable::functionHoistingCandidates const):
(JSC::EvalExecutable::topLevelFunctionDecls const):

Canonical link: <a href="https://commits.webkit.org/295349@main">https://commits.webkit.org/295349@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46829d6b8c478d3135fb82f61cc60d2f2beae8e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104790 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14924 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110004 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55464 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33048 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79568 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107796 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19359 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94571 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59875 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19116 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12648 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54846 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97471 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88814 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12695 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112411 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103408 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31955 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23494 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88648 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32319 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90797 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88275 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33165 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10928 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27268 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17002 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31880 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37234 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31672 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35013 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33231 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/98fd56e1e4de8ea2164d98ff4412e331f9386007. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45816

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/46829d6b8c478d3135fb82f61cc60d2f2beae8e2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45816

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Careful optimization of eval
https://bugs.webkit.org/show_bug.cgi?id=293474
rdar://151902736

Reviewed by Yijia Huang.

This patch carefully restructures executeEval code.

1. Avoid doing some unnecessary things when we do not need to
   instantiate any variables.
2. Use vmEntryToJavaScriptWith0Arguments if necessary.
3. Use ensureStillAliveHere to keep JITCode alive instead of ref-ing.
   This ThreadSafeRefCount turned out super costly.

* Source/JavaScriptCore/bytecode/CodeBlock.h:
(JSC::CodeBlock::functionDecls):
(JSC::CodeBlock::functionExprs):
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.h:
(JSC::UnlinkedCodeBlock::functionDecls const):
(JSC::UnlinkedCodeBlock::functionExprs const):
* Source/JavaScriptCore/bytecode/UnlinkedEvalCodeBlock.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeEval):
* Source/JavaScriptCore/runtime/EvalExecutable.h:
(JSC::EvalExecutable::variables const):
(JSC::EvalExecutable::functionHoistingCandidates const):
(JSC::EvalExecutable::topLevelFunctionDecls const):

Canonical link: https://commits.webkit.org/295349@main
